### PR TITLE
Unit tests mushroom service

### PR DESCRIPTION
**Pull request by bencikpeter**

### CODE REVIEW

File: service/src/test/java/cz/muni/fi/pa165/mushrooms/service/MushroomServiceImplTest.java [hunk 1]
```diff
@@ -0,0 +1,299 @@
...
+    private MushroomDao mushroomDao;
+
+    @Tested(fullyInitialized = true)
+    private MushroomServiceImpl service;
+
+    private class MockDatabase {
+        private Map<Long, Mushroom> database = new HashMap<>();
+        private long databaseCounter = 0;
+
+        private void validate(Mushroom mushroom){
```
Comment (bkompis) [inline service/src/test/java/cz/muni/fi/pa165/mushrooms/service/MushroomServiceImplTest.java]: move to TestUtils?

File: service/src/test/java/cz/muni/fi/pa165/mushrooms/service/MushroomServiceImplTest.java [hunk 2]
```diff
@@ -0,0 +1,299 @@
...
+        public void update(Mushroom mushroom) {
+            validate(mushroom);
+            if (mushroom.getId() == null) throw new IllegalArgumentException("not persisted - cannot be updated");
+            if (database.replace(mushroom.getId(),mushroom) == null) throw new IllegalArgumentException("no object with such id in DB - cannot be updated");;
+        }
+
+        public void delete(Mushroom mushroom){
+            validate(mushroom);
+            if (mushroom.getId() == null) throw new IllegalArgumentException("no id assigned");
+            //TODO: deleting non persisted should be OK or NOK?
```
Comment (bkompis) [inline service/src/test/java/cz/muni/fi/pa165/mushrooms/service/MushroomServiceImplTest.java]: same as in dao layer - should not throw an exception, but should not modify db (if i recall correctly)

File: service/src/test/java/cz/muni/fi/pa165/mushrooms/service/MushroomServiceImplTest.java [hunk 3]
```diff
@@ -0,0 +1,299 @@
...
+        public Mushroom findById(Long id) {
+            if (id == null){
+                throw new IllegalArgumentException("null id");
+            }
+            return database.get(id);
+        }
+
+        public List<Mushroom> findByMushroomType(MushroomType mushroomType){
+            List<Mushroom> typedList = new ArrayList<>();
+            List<Mushroom> dumpedDatabase = Collections.unmodifiableList(new ArrayList<>(database.values()));
```
Comment (bkompis) [inline service/src/test/java/cz/muni/fi/pa165/mushrooms/service/MushroomServiceImplTest.java]: why not iterate over database.values()?

File: service/src/test/java/cz/muni/fi/pa165/mushrooms/service/MushroomServiceImplTest.java [hunk 4]
```diff
@@ -0,0 +1,299 @@
...
+    private static Mushroom setupMushroom(String name, MushroomType type, String fromM, String toM){
+        Mushroom m = new Mushroom();
+        m.setName(name);
+        m.setIntervalOfOccurrence(fromM,toM);
+        m.setType(type);
+        return m;
+    }
+
+    @Before
+    public void Setup(){
```
Comment (bkompis) [inline service/src/test/java/cz/muni/fi/pa165/mushrooms/service/MushroomServiceImplTest.java]: setUp()

File: service/src/test/java/cz/muni/fi/pa165/mushrooms/service/MushroomServiceImplTest.java [hunk 5]
```diff
@@ -0,0 +1,299 @@
...
+        mushroom1 = setupMushroom("some",MushroomType.UNEDIBLE,"june","july");
+        mushroom2 = setupMushroom("other",MushroomType.POISONOUS,"june","july");
+        mushroom3 = setupMushroom("different",MushroomType.UNEDIBLE,"may","september");
+
+
+
+        new Expectations(){{
+            mushroomDao.findById(anyLong);
+            result = new Delegate() {
+                Mushroom foo(Long id){
```
Comment (bkompis) [inline service/src/test/java/cz/muni/fi/pa165/mushrooms/service/MushroomServiceImplTest.java:133]: add error cases to all Delegates

File: service/src/test/java/cz/muni/fi/pa165/mushrooms/service/MushroomServiceImplTest.java [hunk 6]
```diff
@@ -0,0 +1,299 @@
...
+            }; minTimes = 0;
+        }};
+    }
+
+    @Test
+    public void findAllMushrooms(){
+
+        assertThat(service.findAllMushrooms()).isEmpty();
+
+        database.create(mushroom1);
```
Comment (bkompis) [inline service/src/test/java/cz/muni/fi/pa165/mushrooms/service/MushroomServiceImplTest.java:194]: dependent test

File: service/src/test/java/cz/muni/fi/pa165/mushrooms/service/MushroomServiceImplTest.java [hunk 7]
```diff
@@ -0,0 +1,299 @@
...
+    }
+
+    @Test
+    public void FindMushroomByName(){
+
+        database.create(mushroom1);
+        database.create(mushroom2);
+
+        //null name
+        assertThatThrownBy(() -> service.findMushroomByName(null)).isInstanceOf(IllegalArgumentException.class);
```
Comment (bkompis) [inline service/src/test/java/cz/muni/fi/pa165/mushrooms/service/MushroomServiceImplTest.java]: split to separate test methods

**Review by bkompis — CHANGES_REQUESTED** [review]
Dependent test cases, test methods testing several things at once, kinda small number of tests

File: service/src/test/java/cz/muni/fi/pa165/mushrooms/service/MushroomServiceImplTest.java [hunk 8]
```diff
@@ -210,7 +212,7 @@ public void findMushroomById(){
         database.create(mushroom1);
         database.create(mushroom2);
         //null
-        assertThatThrownBy(() -> service.findMushroomById(null)).isInstanceOf(IllegalArgumentException.class);
+        assertThatThrownBy(() -> service.findMushroomById(null)).isInstanceOf(EntityFindServiceException.class);
```
Comment (bkompis) [inline service/src/test/java/cz/muni/fi/pa165/mushrooms/service/MushroomServiceImplTest.java]: DataAccessException

File: service/src/test/java/cz/muni/fi/pa165/mushrooms/service/MushroomServiceImplTest.java (hunk 6 above)
Comment (bencikpeter) [reply service/src/test/java/cz/muni/fi/pa165/mushrooms/service/MushroomServiceImplTest.java:194]: not really... tested entity is service, database is expected to be reliable (is a mock object)

File: service/src/test/java/cz/muni/fi/pa165/mushrooms/service/MushroomServiceImplTest.java [hunk 9]
```diff
@@ -208,94 +199,208 @@ public void findAllMushrooms(){
...
 
     @Test
-    public void createMushroom(){
+    public void createMushroom_valid(){
 
         assertThat(mushroom1.getId()).isNull();
         service.createMushroom(mushroom1);
         assertThat(mushroom1.getId()).isNotNull();
-
+        assertThat(database.findById(mushroom1.getId())).isEqualToComparingFieldByField(mushroom1);
```
Comment (bkompis) [inline service/src/test/java/cz/muni/fi/pa165/mushrooms/service/MushroomServiceImplTest.java]: should use business equals -.-

**Review by bkompis — CHANGES_REQUESTED** [review]
one incorrect equals use, otherwise ok

File: service/src/test/java/cz/muni/fi/pa165/mushrooms/service/MushroomServiceImplTest.java (hunk 6 above)
Comment (bkompis) [reply service/src/test/java/cz/muni/fi/pa165/mushrooms/service/MushroomServiceImplTest.java:194]: oh, my bad - reviewing too fast..

**Review by bkompis — APPROVED** [review]

**Pull request merged**